🎯Goal  

To become confident in core machine learning techniques and be able to give correct reasoning for my choices 

🛠️Setting up the enviorment  

First, we import the necessary libraries.

**scikit-learn**: for ml models  
**Pandas & NumPy**: For data manipulation  
**matplotlib**: For visulization📊  


In [2]:
#Importing required library
import numpy as np 
import pandas as pd
import sklearn
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

In [3]:
#Loading the csv
heart = pd.read_csv("./data/heart.csv")

In [3]:
#top values of dataset
heart.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

In [5]:
#shape of dataset
heart.shape

(1025, 14)

In [6]:
heart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


In the output of the above code we can see that there are no null or missing values so we dont have to handle them 

We also do not have any categorical value so we have no nned to encode it

In [9]:
heart.describe()

age          sex           cp     trestbps        chol  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.00000   
mean     54.434146     0.695610     0.942439   131.611707   246.00000   
std       9.072290     0.460373     1.029641    17.516718    51.59251   
min      29.000000     0.000000     0.000000    94.000000   126.00000   
25%      48.000000     0.000000     0.000000   120.000000   211.00000   
50%      56.000000     1.000000     1.000000   130.000000   240.00000   
75%      61.000000     1.000000     2.000000   140.000000   275.00000   
max      77.000000     1.000000     3.000000   200.000000   564.00000   

               fbs      restecg      thalach        exang      oldpeak  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.000000   
mean      0.149268     0.529756   149.114146     0.336585     1.071512   
std       0.356527     0.527878    23.005724     0.472772     1.175053   
min       0.000000     0.000000    71.000000     0.000000     0.000000   
25%       0.000000     0.000000   132.000000     0.000000     0.000000   
50%       0.000000     1.000000   152.000000     0.000000     0.800000   
75%       0.000000     1.000000   166.000000     1.000000     1.800000   
max       1.000000     2.000000   202.000000     1.000000     6.200000   

             slope           ca         thal       target  
count  1025.000000  1025.000000  1025.000000  1025.000000  
mean      1.385366     0.754146     2.323902     0.513171  
std       0.617755     1.030798     0.620660     0.500070  
min       0.000000     0.000000     0.000000     0.000000  
25%       1.000000     0.000000     2.000000     0.000000  
50%       1.000000     0.000000     2.000000     1.000000  
75%       2.000000     1.000000     3.000000     1.000000  
max       2.000000     4.000000     3.000000     1.000000

What does this dataset represent??  

This data set dates from 1988 and consists of four databases: Cleveland, Hungary, Switzerland, and Long Beach V. It contains 76 attributes, including the predicted attribute, but all published experiments refer to using a subset of 14 of them. The "target" field refers to the presence of heart disease in the patient. It is integer valued 0 = no disease and 1 = disease.  

- In the age column 1 = male and 0 = female
- cp stands for chest pain type (4 values)
- trestbps means resting blood pressure (in mm Hg on admission to the hospital)
- chol is the serum cholestoral in mg/dl
- fbs stands for fasting blood sugar &gt; 120 mg/dl (1 = true; 0 = false)
- restecg means resting electrocardiographic results
- thalach is the maximum heart rate achieved
- exang is exercise induced angina (1 = yes; 0 = no)
- oldpeak = ST depression induced by exercise relative to rest
- slope is the slope of the peak exercise ST segment
- ca is the number of major vessels (0-3) colored by flourosopy
- thal: 0 = normal; 1 = fixed defect; 2 = reversable defect



In [4]:
#spliting features
X = heart.drop('target', axis=1)
Y = heart['target']

In [5]:
#creating train and test set
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(
    X,Y,
    test_size=0.2,
    random_state=42,
    stratify=Y
)

random_state = sets a fixed random seed which gurantees the same split everytime the code is executed  

stratify = this defines that both traning and test set have smae class distribution

In [ ]:
#Logistic regression model
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

log_model = Pipeline([
    ('scaler', StandardScaler()),
    ('model', LogisticRegression(
        max_iter=2000,
        solver='lbfgs'
    ))
])

log_model.fit(X_train, Y_train)
log_preds = log_model.predict(X_test)


Logistic regression is Scale-sensitive and hence we had to scale the columns like age, chol, trestbps, etc..  

The reason to use pipline instead of not doing the scaling manually is that pipeline is safer, cleaner, and less error-prone.

In [ ]:
#Decision Tree Model
from sklearn.tree import DecisionTreeClassifier

dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, Y_train)

dt_preds = dt_model.predict(X_test)


In [11]:
#Random forest model
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(
    n_estimators=200,
    random_state=42
)
rf_model.fit(X_train, Y_train)

rf_preds = rf_model.predict(X_test)

In our random forest we have 200 trees, which is reasonable for our task.

Model Evaluation  

After traing the models, their performance is measured on unseen test data.  
All models are evaluated using same test set and the same metrics.

Evaluation metric used  

1. Accuracy  

Accuracy measures the proportion of correct predictions  
$$
Accuracy=\frac{Correct Predictions}{Total Predictions}
$$
It provides a quick overall performance measure  
<br>
<br>

2. Confusion Matrix  

The confusion matrix shows:
- True Positives
- True Negatives
- False Positives
- False Negatives

This helps understand where the model makes mistakes, which is especially important in medical classification tasks.
<br>
<br>

3. Precison and Recall

Since this is a classification problem, precision and recall are used.
- **Precision**: How many predicted positives are actually correct
- **Recall**: How many actual positives are correctly identified

In [12]:
# Logistic Regression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

log_accuracy = accuracy_score(Y_test, log_preds)
log_cm = confusion_matrix(Y_test, log_preds)

print("Logistic Regression Accuracy:", log_accuracy)
print("Confusion Matrix:\n", log_cm)
print(classification_report(Y_test, log_preds))

Logistic Regression Accuracy: 0.8097560975609757
Confusion Matrix:
 [[70 30]
 [ 9 96]]
              precision    recall  f1-score   support

           0       0.89      0.70      0.78       100
           1       0.76      0.91      0.83       105

    accuracy                           0.81       205
   macro avg       0.82      0.81      0.81       205
weighted avg       0.82      0.81      0.81       205



In [13]:
#Decision Tree
dt_accuracy = accuracy_score(Y_test, dt_preds)
dt_cm = confusion_matrix(Y_test, dt_preds)

print("Decision Tree Accuracy:", dt_accuracy)
print("Confusion Matrix:\n", dt_cm)
print(classification_report(Y_test, dt_preds))

Decision Tree Accuracy: 0.9853658536585366
Confusion Matrix:
 [[100   0]
 [  3 102]]
              precision    recall  f1-score   support

           0       0.97      1.00      0.99       100
           1       1.00      0.97      0.99       105

    accuracy                           0.99       205
   macro avg       0.99      0.99      0.99       205
weighted avg       0.99      0.99      0.99       205



In [15]:
#Random Forest
rf_accuracy= accuracy_score(Y_test, rf_preds)
rf_cm = confusion_matrix(Y_test, rf_preds)

print("Random Forest Accuracy:", rf_accuracy)
print("Confusion Matrix:\n", rf_cm)
print(classification_report(Y_test, rf_preds))

Random Forest Accuracy: 1.0
Confusion Matrix:
 [[100   0]
 [  0 105]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       100
           1       1.00      1.00      1.00       105

    accuracy                           1.00       205
   macro avg       1.00      1.00      1.00       205
weighted avg       1.00      1.00      1.00       205



In [17]:
#Model Comparison Table
import pandas as pd

comparison_table = pd.DataFrame({
    'Model': ['Logistic Regression', 'Decision Tree', 'Random Forest'],
    'Accuracy': [log_accuracy, dt_accuracy, rf_accuracy]
})

comparison_table


Model  Accuracy
0  Logistic Regression  0.809756
1        Decision Tree  0.985366
2        Random Forest  1.000000

Analysis  


Which model performed best?  
- The Random Forest model performed the best among the three models, as it achieved 100% accuracy, precision, and recall, indicating perfect classification on the dataset. 
- Logistic Regression was the weakest model, as it exhibited a low recall (70%), meaning it missed a significant number of positive cases (high false negatives).
- The Decision Tree also performed strongly, achieving 100% recall and only a small number of false positives, resulting in high overall performance, though not as perfect as the Random Forest.


Why do you think that happened?
- Logistic regression performed the worst because it assumes a liner relationship between features and outcome. If the true relationship is non-linear the linear regression struggles, leading to missed postives cases.
- Random forest perforemed perfectly because its an ensemble method that combines many decision tree(200 in our case). Then averages their predictions which reduces overfitting and varience, casuing it to easily capture non-linear relation with complex boundries. Since the data is clean and small the model had a perfect performance.
- Decision tree also performed very well as tree can easily fit to complex decison boundries and learn from traning patterns. However, a sing tree is prone to overfiiting and hence few false positives compared to random forest.


Did any model overfit?
- Yes the random forest shows perfect performance which indicates overfitting or data leakage.

What would you try next?
- Training with a larget dataset or applying cross-validation to current models.